KAFKA-14129: KRaft must check manual assignments for createTopics are contiguous

cmccabe opened the pull request:
KRaft should validate that manual assignments given to createTopics are contiguous. In other words,
they must start with partition 0, and progress through 1, 2, 3, etc. ZK mode does this, but KRaft
mode previously did not. Also fix a null pointer exception when the placement for partition 0
was not specified.

Convert over AddPartitionsTest to use KRaft. This PR converts all of the test except for some of
the placement logic tests, which will need to be redone for KRaft mode in a future change.

Fix null pointer exception in KRaftMetadataCache#getPartitionInfo.  Specifically, we should not
assume that the partition will be found in the hash map. This is another case where we had
"Some(x)" but it should be "Option(x)."

BrokerLifecycleManager should be initialized in the constructor of BrokerServer. Otherwise there
will be a null pointer exception if BrokerServer#brokerState is called prior to BrokerServer#start.

hachikuji reviewed (approved):
Good find. LGTM.

dengziming commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -81,6 +81,8 @@ class BrokerServer(
   val controllerQuorumVotersFuture: CompletableFuture[util.Map[Integer, AddressSpec]]
 ) extends KafkaBroker {
 
+  val lifecycleManager: BrokerLifecycleManager = new BrokerLifecycleManager(config, time, threadNamePrefix)
```
This change will make some tests fail since `lifecycleManager` is stateful(for example, initialCatchUpFuture ) but we are stopping and restarting it in `LeaderElectionCommandTest`.

cmccabe replied:
thanks for the heads up. I will go with a different fix.

cmccabe committed:
KAFKA-14129: KRaft must check manual assignments for createTopics are contiguous

KRaft should validate that manual assignments given to createTopics are contiguous. In other words,
they must start with partition 0, and progress through 1, 2, 3, etc. ZK mode does this, but KRaft
mode previously did not. Also fix a null pointer exception when the placement for partition 0
was not specified.

Convert over AddPartitionsTest to use KRaft. This PR converts all of the test except for some of
the placement logic tests, which will need to be redone for KRaft mode in a future change.

Fix null pointer exception in KRaftMetadataCache#getPartitionInfo.  Specifically, we should not
assume that the partition will be found in the hash map. This is another case where we had
"Some(x)" but it should be "Option(x)."

Fix a potential null pointer exception in BrokerServer#state.
